runtime/j9vm/jvm.c

```diff
@@ -1218,73 +1226,156 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, DIR_SEPARATOR_STR "classlib.properties");
+	propsFile = props_file_open(PORTLIB, jvmBufferData(propsPathBuffer), NULL, 0);
+	free(propsPathBuffer);
+	propsPathBuffer = NULL;
+
+	if (NULL != propsFile) {
+		const char *shape = NULL;
+		const char *version = NULL;
+		UDATA decoded = 0;
+		
+		shape = props_file_get(propsFile, "shape");
 		if (NULL == shape) {
-	#ifdef DEBUG
-		printf("No 'shape' property in %s\n", classlibProps);
-	#endif
+#ifdef DEBUG
+			printf("No 'shape' property in %s\n", propsFile);
+#endif
 			goto bail;
 		}
 
-		version = props_file_get(classlibProps, "version");
+		version = props_file_get(propsFile, "version");
 		if (NULL == version) {
-	#ifdef DEBUG
-		printf("No 'version' property in %s\n", classlibProps);
-	#endif
+#ifdef DEBUG
+			printf("No 'version' property in %s\n", propsFile);
+#endif
 			goto bail;
 		}
 	
 		decoded = decodeSetting("shape", shape, SHAPE_SETTINGS, NUM_SHAPE_SETTINGS);
-		if (decoded == 0) {
+		if (0 == decoded) {
 			goto bail;
 		}
 		finalVersion |= decoded;
 	
 		decoded = decodeSetting("version", version, VERSION_SETTINGS, NUM_VERSION_SETTINGS);
-		if (decoded == 0) {
+		if (0 == decoded) {
 			goto bail;
 		}
 		finalVersion |= decoded;
 
-	bail:
-		props_file_close(classlibProps);
+bail:
+		props_file_close(propsFile);
+	} else {
+#ifdef DEBUG
+		printf("Could not open %s\n", propsFile);
+#endif
+	}
+	
+	return finalVersion;
+}
+
+/**
+ * Attempt loading 'release' file, and get Java version info.
+ * If the file is found, 'JAVA_VERSION' value is retrieved and decoded as following:
+ * "1.8.0_xxx"	--- Java 8, 'J2SE_18 | J2SE_SHAPE_SUN';
+ * "9"			--- Java 9, 'J2SE_19 | J2SE_SHAPE_B165';
+ * "10"			--- Java 18.3, 'J2SE_2018_3 | J2SE_SHAPE_B1803';
+ * Others		--- Latest Java, 'J2SE_LATEST | J2SE_SHAPE_LATEST'.
+ * Note: 'release' file contains JAVA_VERSION="10" for Java 18.3 at this moment.
+ * Otherwise, 0 is returned.
+ *
+ * @return 	'J2SE_18 | J2SE_SHAPE_SUN', 'J2SE_19 | J2SE_SHAPE_B165',
+ *			'J2SE_2018_3 | J2SE_SHAPE_B1803', 'J2SE_LATEST | J2SE_SHAPE_LATEST'
+ *			according to the 'JAVA_VERSION' value found in 'release';
+ *			or 0 if otherwise.
+ */
+static UDATA
+getVersionFromReleaseFile(void)
+{
+	PORT_ACCESS_FROM_PORT(&j9portLibrary);
+	J9StringBuffer *propsPathBuffer = NULL;
+	j9props_file_t propsFile = NULL;
+	UDATA finalVersion = 0;
+
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, jvmBufferData(j9Buffer));
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, DIR_SEPARATOR_STR "release");
+	propsFile = props_file_open(PORTLIB, jvmBufferData(propsPathBuffer), NULL, 0);
+	free(propsPathBuffer);
+	propsPathBuffer = NULL;
+	if (NULL != propsFile) {
+		const char *version = props_file_get(propsFile, "JAVA_VERSION");
+		if (NULL != version) {
+#define		JAVA_VERSION_8		"\"1.8.0"		/* its usual format is "1.8.0_xxx" */
+			if (!strncmp(version, JAVA_VERSION_8, strlen(JAVA_VERSION_8))) {
```

Comment:
Using `sizeof(JAVA_VERSION_8) - 1` avoids the runtime function call to `strlen`.

Reply:
Fixed. Pls have another look.